### ISSUE TITLE
CMS - Lighthouse: push service locations

### ISSUE THREAD

**Issue by swirtSJW** [department-of-veterans-affairs/va.gov-cms#8017]
## User Story or Problem Statement

As a lighthouse Facility API data maintainer, I need to have  the location of facility services provided with the service when pushed to us. Initial discussion with Lighthouse in Slack](https://dsva.slack.com/archives/C02BTJTDFTN/p1646760404599479)

~This is likely blocked by https://github.com/department-of-veterans-affairs/va.gov-cms/issues/6931.  The only reason it would not be blocked is if we don't send hours for now.~  That was resolved.

Service location is a paragraph of paragraphs.

### Current Model
```
{
  "detailed_services": [
    {
      "name": "Audiology and speech",
      "active": true,
      "description_national": "<p>We diagnose and treat conditions...</p>\r\n",
      "description_system": "<h3>Audiology at VA...</h3>",
      "service_api_id": "audiology",
      "appointment_leadin": "<p><em>Contact us to schedule, reschedule, or cancel your appointment. If a referral is required, you'll need to contact your primary care provider first.</em></p>\r\n",
      "appointment_phones": [
        {
          "type": "tel",
          "label": "Main phone",
          "number": "800-329-8387",
          "extension": null
        }
      ],
      "referral_required": "true",
      "walk_ins_accepted": "false"
    }
  ]
}
```

### Proposed Model
```
{
  "detailed_services": [
    {
      "name": "Audiology and speech",
      "active": true,
      "description_national": "<p>We diagnose and treat conditions...</p>\r\n",
      "description_system": "<h3>Audiology at VA...</h3>",
      "service_api_id": "audiology",
      "appointment_leadin": "<p>Contact us to schedule, reschedule, or cancel your appointment. If a referral is required, you'll need to contact your primary care provider first.</em></p>",
      "appointment_phones": [
        {
          "type": "tel",
          "label": "Main phone",
          "number": "800-329-8387",
          "extension": null
        }
      ],
      "referral_required": true, // Changed to boolean value in 4769
      "service_locations": [ // Start of new service_locations array - may contain multiple values
        {
          "office_name": "ENT Clinic", // Can be null if no value is provided as the field_clinic_name (name of office or location)
          "email_contacts:": [
            {
              "email_address": "georgea@va.gov",
              "email_label": "George Anderson"
            },
            {
              "email_address": "confirmations@va.gov",
              "email_label": "Confirm your appointment"
            }
          ],
          "facility_service_hours": [
            {
              "Monday": "9:00 am - 5:00 pm",
              "Tuesday": "9:00 am - 4:00 pm",
              "Wednesday": "9:00 am - 4:00 pm  Call for apt",
              "Thursday": "9:00 am - 4:00 pm",
              "Friday": "By appointment only",
              "Saturday": "Closed",
              "Sunday": "Closed"
            }
          ],
          "additional_hours_info": "Please use call for an apt outside...",
          "phones": [
            {
              "extension": "123",
              "label": "Appointment phone",
              "number": "555-555-1212",
              "type": "tel"
            },
            {
              "label": "TTY",
              "number": "222-222-1212",
              "type": "tty"
            }
          ],
          "service_location_address": {
            "address_line1": "122 Main St.",
            "address_line2": null,
            "state": "NY",
            "building_name_number": "Baxter Bulding",
            "clinic_name": "Baxter Clinic",
            "country_code": "US",
            "city": "Rochester",
            "zip_code": "14623-1345",
            "wing_floor_or_room_number": "Wing East"
          }
        },
      ],
      "walk_ins_accepted": false // Changed to boolean value in 4769
    }
  ]
}
```

### Sprint 90 commitments - 3 points
- [ ] Run the bulk push of all services
- [ ] Provide report of all services sent, to Lighthouse

### For close
- [ ] On service push, the service location(s) appear in the payload.
- [ ] In the event there are no service locations, use facility address.
- [ ] In the event that there are service locations, if any are checked Use facility address, include that address.
- [ ] Each service location contains the data including: address, hours and contact info (see data model above)
![image](https://user-images.githubusercontent.com/5752113/157305559-f4037797-f78c-441b-9c15-d38427e5e813.png)
- [ ] empty fields must be included in the push and have a value of NULL.
- [ ] keyed as `locations` and must be an array of location arrays.
- [ ] fields that are being pushed have CM Documents created that reflect they are being pushed to Facility API.

## Implementation Recs
- Push code lives here docroot/modules/custom/va_gov_post_api/src/Service/PostFacilityService.php  within the call to getPayload()
   -  addition should look like `$service->locations = $this->getLocations();`
- field names in the push should ideally match drupal field names without the `field_` preface.  
- field names once assembled should be coordinated approved by lighthouse team (sometimes CMS field names are gibberish and make no sense outside the context of drupal)  Keep deviations to a minimum, but do what has to be done to keep them meaningful to Lighthouse who will be surfacing them to the public api.

## Design principles

Veteran-centered
- [x] `Single source of truth`: Increase reliability and consistency of content on VA.gov by providing a single source of truth.
- [ ] `Accessible, plain language`: Provide guardrails and guidelines to ensure content quality.
- [ ] `Purposely structured content`: Ensure Content API can deliver content whose meaning matches its structure.
- [ ] `Content lifecycle governance`: Produce tools, processes and policies to maintain content quality throughout its lifecycle.

Editor-centered
- [ ] `Purpose-driven`: Create an opportunity to involve the editor community in VA’s mission and content strategy goals.
- [ ] `Efficient`: Remove distractions and create clear, straightforward paths to get the job done.
- [ ] `Approachable`: Offer friendly guidance over authoritative instruction.
- [ ] `Consistent`: Reduce user’s mental load by allowing them to fall back on pattern recognition to complete tasks.
- [ ] `Empowering`: Provide clear information to help editors make decisions about their work.

**Comment by JayDarnell** [comment]
Work can begin on this item once V1 is live.

**Comment by JayDarnell** [comment]
Updated discussion on proposed data model: https://dsva.slack.com/archives/C02BTJTDFTN/p1664375519019619

**Comment by JayDarnell** [comment]
```
{
    "detailed_services": [
      {
        "name": "Audiology and speech",
        "active": true,
        "description_national": "<p>We diagnose and treat conditions...</p>\r\n",
        "description_system": "<h3>Audiology at VA...</h3>",
        "service_api_id": "audiology",
        "appointment_leadin": "<p>Contact us to schedule, reschedule, or cancel your appointment. If a referral is required, you'll need to contact your primary care provider first.</em></p>",
        "appointment_phones": [
          {
            "type": "tel",
            "label": "Main phone",
            "number": "800-329-8387",
            "extension": null
          }
        ],
        "referral_required": "true",
        "service_locations": [
          {
            "email_contacts:": [
              {
                "email_address": "georgea@va.gov",
                "email_label": "George Anderson"
              },
              {
                "email_address": "confirmations@va.gov",
                "email_label": "Confirm your appointment"
              }
            ],
            "facility_service_hours": [
              {
                "Monday": "9-5",
                "Tuesday": "9-4",
                "Wednesday": "9-4  Call for apt",
                "Thursday": "9-4",
                "Friday": "By appointment only",
                "Saturday": "Closed",
                "Sunday": "Closed"
              }
            ],
            "additional_hours_info": "Please use call for an apt outside...",
            "phones": [
              {
                "extension": "123",
                "label": "Appointment phone",
                "number": "555-555-1212",
                "type": "tel",
              },
              {
                "label": "TTY",
                "number": "222-222-1212",
                "type": "tty",
              }
            ],
            "service_location_address": {
              "address_line1": "122 Main St.",
              "address_line2": null,
              "state": "NY",
              "building_name_number": "Baxter Bulding",
              "clinic_name": "Baxter Clinic",
              "country_code": "US",
              "city": "Rochester",
              "zip_code": "14623-1345",
              "wing_floor_or_room_number": "Wing East",
            },
          }
        ],
        "walk_ins_accepted": "false"
      }
    ]
  }
```

**Comment by JayDarnell** [comment]
Example of proposed data model (with two service locations): 
```
{
  "detailed_services": [
    {
      "name": "Audiology and speech",
      "active": true,
      "description_national": "<p>We diagnose and treat conditions...</p>\r\n",
      "description_system": "<h3>Audiology at VA...</h3>",
      "service_api_id": "audiology",
      "appointment_leadin": "<p>Contact us to schedule, reschedule, or cancel your appointment. If a referral is required, you'll need to contact your primary care provider first.</em></p>",
      "appointment_phones": [
        {
          "type": "tel",
          "label": "Main phone",
          "number": "800-329-8387",
          "extension": null
        }
      ],
      "referral_required": "true",
      "service_locations": [
        {
          "office_name": "ENT Clinic",
          "email_contacts:": [
            {
              "email_address": "georgea@va.gov",
              "email_label": "George Anderson"
            },
            {
              "email_address": "confirmations@va.gov",
              "email_label": "Confirm your appointment"
            }
          ],
          "facility_service_hours": [
            {
              "Monday": "9:00 am - 5:00 pm",
              "Tuesday": "9:00 am - 4:00 pm",
              "Wednesday": "9:00 am - 4:00 pm  Call for apt",
              "Thursday": "9:00 am - 4:00 pm",
              "Friday": "By appointment only",
              "Saturday": "Closed",
              "Sunday": "Closed"
            }
          ],
          "additional_hours_info": "Please use call for an apt outside...",
          "phones": [
            {
              "extension": "123",
              "label": "Appointment phone",
              "number": "555-555-1212",
              "type": "tel"
            },
            {
              "label": "TTY",
              "number": "222-222-1212",
              "type": "tty"
            }
          ],
          "service_location_address": {
            "address_line1": "122 Main St.",
            "address_line2": null,
            "state": "NY",
            "building_name_number": "Baxter Bulding",
            "clinic_name": "Baxter Clinic",
            "country_code": "US",
            "city": "Rochester",
            "zip_code": "14623-1345",
            "wing_floor_or_room_number": "Wing East"
          }
        },
        {
          "office_name": null,
          "email_contacts:": [
            {
              "email_address": "jamesd@va.gov",
              "email_label": "James Dawson"
            }
          ],
          "facility_service_hours": [
            {
              "Monday": "9:00 am - 5:00 pm",
              "Tuesday": "9:00 am - 5:00 pm",
              "Wednesday": "9:00 am - 5:00 pm",
              "Thursday": "9:00 am - 5:00 pm",
              "Friday": "By appointment only",
              "Saturday": "Closed",
              "Sunday": "Closed"
            }
          ],
          "additional_hours_info": null,
          "phones": [
            {
              "extension": "123",
              "label": "Appointment phone",
              "number": "555-555-2323",
              "type": "tel"
            }
          ],
          "service_location_address": {
            "address_line1": "152 Dufferin St.",
            "address_line2": null,
            "state": "NY",
            "building_name_number": null,
            "clinic_name": null,
            "country_code": "US",
            "city": "Rochester",
            "zip_code": "14623-1345",
            "wing_floor_or_room_number": null
          }
        }
      ],
      "walk_ins_accepted": "false"
    }
  ]
}
```

**Comment by swirtSJW** [comment]
Finalized data
```
{
  "detailed_services": [
    {
      "name": "Audiology and speech",
      "active": true,
      "description_national": "<p>We diagnose and treat conditions...</p>\r\n",
      "description_system": "<h3>Audiology at VA...</h3>",
      "service_api_id": "audiology",
      "appointment_leadin": "<p>Contact us to schedule, reschedule, or cancel your appointment. If a referral is required, you'll need to contact your primary care provider first.</em></p>",
      "appointment_phones": [
        {
          "type": "tel",
          "label": "Main phone",
          "number": "800-329-8387",
          "extension": null
        }
      ],
      "referral_required": true,
      "service_locations": [
        {
          "office_name": "ENT Clinic",
          "email_contacts:": [
            {
              "email_address": "georgea@va.gov",
              "email_label": "George Anderson"
            },
            {
              "email_address": "confirmations@va.gov",
              "email_label": "Confirm your appointment"
            }
          ],
          "facility_service_hours": [
            {
              "Monday": "9:00 am - 5:00 pm",
              "Tuesday": "9:00 am - 4:00 pm",
              "Wednesday": "9:00 am - 4:00 pm  Call for apt",
              "Thursday": "9:00 am - 4:00 pm",
              "Friday": "By appointment only",
              "Saturday": "Closed",
              "Sunday": "Closed"
            }
          ],
          "additional_hours_info": "Please use call for an apt outside...",
          "phones": [
            {
              "extension": "123",
              "label": "Appointment phone",
              "number": "555-555-1212",
              "type": "tel"
            },
            {
              "label": "TTY",
              "number": "222-222-1212",
              "type": "tty"
            }
          ],
          "service_location_address": {
            "address_line1": "122 Main St.",
            "address_line2": null,
            "state": "NY",
            "building_name_number": "Baxter Bulding",
            "clinic_name": "Baxter Clinic",
            "country_code": "US",
            "city": "Rochester",
            "zip_code": "14623-1345",
            "wing_floor_or_room_number": "Wing East"
          }
        },
        {
          "office_name": null,
          "email_contacts:": [
            {
              "email_address": "jamesd@va.gov",
              "email_label": "James Dawson"
            }
          ],
          "facility_service_hours": [
            {
              "Monday": "9:00 am - 5:00 pm",
              "Tuesday": "9:00 am - 5:00 pm",
              "Wednesday": "9:00 am - 5:00 pm",
              "Thursday": "9:00 am - 5:00 pm",
              "Friday": "By appointment only",
              "Saturday": "Closed",
              "Sunday": "Closed"
            }
          ],
          "additional_hours_info": null,
          "phones": [
            {
              "extension": "123",
              "label": "Appointment phone",
              "number": "555-555-2323",
              "type": "tel"
            }
          ],
          "service_location_address": {
            "address_line1": "152 Dufferin St.",
            "address_line2": null,
            "state": "NY",
            "building_name_number": null,
            "clinic_name": null,
            "country_code": "US",
            "city": "Rochester",
            "zip_code": "14623-1345",
            "wing_floor_or_room_number": null
          }
        }
      ],
      "walk_ins_accepted": false
    }
  ]
}

```

**Comment by swirtSJW** [comment]
@kmariepat-cityfriends this seems defined enough now that it is unblocked and we can add it to an upcoming sprint.

**Comment by kmariepat-cityfriends** [comment]
from lighthouse meeting, need to add sub-name

**Comment by swirtSJW** [comment]
Validating as I go
- [ ] $service->name
- [ ]  $service->active
- [ ] $service->description_national
- [ ]  $service->description_system 
- [ ] $service->service_api_id 
- [ ] $service->appointment_leadin 
- [ ]  $service->appointment_phones 
- [ ]   $service->referral_required 
- [ ]  $service->service_locations 
- [ ]    $service->walk_ins_accepted

**Comment by swirtSJW** [comment]
Agreed to move the three walking, reservations and __ fields to be inside the service location since it will be part of that in the nearish future https://dsva.slack.com/archives/C02BTJTDFTN/p1682419182113469

**Comment by swirtSJW** [comment]
Also agreed yesterday on the call that we would go back to sending stings and not pure booleans for those fields 
https://dsva.slack.com/archives/C02BTJTDFTN/p1682341200384759

**Comment by swirtSJW** [comment]
I raised with Lighthouse ([slack thread](https://dsva.slack.com/archives/C02BTJTDFTN/p1682533474365839))that we actually have the same value appearing in the model twice under two different names.  Waiting on feedback from Lighthouse to determine which should be removed 
Model proposed:
![image](https://user-images.githubusercontent.com/5752113/234735553-430e83ca-85f2-4def-9715-1f59690c857e.png)
Reality:
![image](https://user-images.githubusercontent.com/5752113/234735602-a1fe7f90-51bf-4464-975d-5537ccffac46.png)

**Comment by jilladams** [comment]
@swirtSJW I missed the notes on this from scrum - PR merged, but I assume we're still blocked from closing on something from Lighthouse?

**Comment by swirtSJW** [comment]
@jilladams 
Lighthouse sandbox it ready to receive service locations, but they have not move their code to their prod yet.  We can not merge our PR until their code lands on prod.

**Comment by swirtSJW** [comment]
@jilladams Also the confusion of " PR merged"   Our PR that was merged was related to this (had some underlying code) but was not the full code.

- Underlying code https://github.com/department-of-veterans-affairs/va.gov-cms/pull/13776
- Actual code https://github.com/department-of-veterans-affairs/va.gov-cms/pull/13464

**Comment by jilladams** [comment]
6/5 Lighthouse meeting notes: 
* Tested last week; Adam gave the go ahead, next steps waiting on roll out, facilities can merge after 
* Unsure of when the rollout with be; currently locked out of AWS account

Will continue in S86.

**Comment by swirtSJW** [comment]
https://github.com/department-of-veterans-affairs/va.gov-cms/pull/13464#issuecomment-1604489730

**Comment by jilladams** [comment]
From planning: 
* need to progress this, defined S90 ACs
* If LH identifies the problem as a CMS side problem, we will cut a new ticket to track that issue / fix, which would block

**Comment by swirtSJW** [comment]
This code went to prod today.
